docs/creating-nodes/testing.md

```diff
@@ -0,0 +1,271 @@
+---
+layout: docs
+toc: creating-nodes-toc.html
+title: Testing
+---
+
+This guideline explains how to write and execute test code for node.
+
+Because nodes run in Node-RED, you cannot execute the processing of individual units of HTML or JavaScript to test processing.
+In addition, if you use manual testing, you need to stop and restart Node-RED after each test so that cleanup processing takes place.
+This makes testing quite time-consuming.
+Therefore, a tool called [Node Test Helper](https://github.com/node-red/node-red-node-test-helper) (hereafter "helper") has been developed for Node-RED.
+This tool provides a framework for node testing.
+By using this framework, you can easily run nodes at the code level.
+
+Node-RED uses the following libraries for testing, and we recommend their use when using the test code.
+You will need to have an understanding of the specifications of these libraries when creating test code.
+Read the documentation of each library and learn how to use it.
+
+| No. | Library | Purpose | URL |
+| --- | --- | ---| --- |
+| 1 | Mocha | Test framework | [https://mochajs.org](https://mochajs.org/) |
+| 2 | Should.js | Assertion | [https://shouldjs.github.io](https://shouldjs.github.io/) |
+| 3 | Sinon.js | Test double | [http://sinonjs.org](http://sinonjs.org/) |
+| 4 | Nock | Test double | [https://github.com/nock/nock](https://github.com/nock/nock) |
+| 5 | Istanbul | Coverage | [https://github.com/gotwarlost/istanbul](https://github.com/gotwarlost/istanbul) |
+
+### Creating test code
+
+Creating test code consists of the following steps:
+
+1. Preparing the environment
+2. Creating the test code
+3. Running the test code
+
+### 1. Preparing the environment
+
+Use `npm` to install the modules required for testing in the directory in which you are developing your custom node.
+
+```bash
+npm install node-red-node-test-helper should mocha --save-dev
+```
+
+Install modules like `sinon` and `nock` as needed.
+
+### 2. Creating the test code
+
+Create a spec file that contains the test directory and the test code.
+You would typically create a `test` directory in the same directory as package.json, and use the naming rule `file-to-be-tested_spec.js` for the test files themselves.
+[Example]
+
+```text
+.
+├── test-node.js
+├── test-node.html
+├── package.json
+└── test
```

Comment:
I must admit we don't currently do this for most nodes - This is sensible.
(If this is done correctly can we then add a line to run the test via npm in the package.json? or does that then require we have all the deps in every node (inc node-red...?))

Reply:
There is a general discussion over whether a module should include its test material when published to npm. In the wider node community it is a similar discussion to tabs vs spaces - ie, no clear winner and a lot of energy spent not coming to agreement.